### PR TITLE
fix(Middlewares): catch errors in non-global middlewares

### PR DESCRIPTION
**Pull request by NoNameProvided**
This PR fixes an issue with Unhandled promise rejection in non-global middlewares.

### CODE REVIEW

**Comment by NoNameProvided** [comment]
Hmm but how run tests? 

Running simply `npm run test` seems not to work, it runs zero test, and the output is:

```bash
> gulp tests

[20:38:53] Requiring external module ts-node/register
[20:38:53] Using gulpfile ~/budacode/routing-controllers/gulpfile.ts
[20:38:53] Starting 'tests'...
[20:38:53] Starting 'compile'...
[20:38:53] Finished 'compile' after 19 ms
[20:38:53] Starting 'coveragePre'...
[20:38:53] Finished 'coveragePre' after 7.66 ms
[20:38:53] Starting 'coveragePost'...


  0 passing (1ms)

----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
----------|----------|----------|----------|----------|----------------|
All files |      100 |      100 |      100 |      100 |                |
----------|----------|----------|----------|----------|----------------|


=============================== Coverage summary ===============================
Statements   : 100% ( 0/0 )
Branches     : 100% ( 0/0 )
Functions    : 100% ( 0/0 )
Lines        : 100% ( 0/0 )
================================================================================
[20:38:53] Finished 'coveragePost' after 85 ms
[20:38:53] Starting 'coverageRemap'...
[20:38:53] Finished 'coverageRemap' after 18 ms
[20:38:53] Starting 'tslint'...
[20:38:55] Finished 'tslint' after 1.28 s
[20:38:55] Finished 'tests' after 1.42 s
```

**Comment by 19majkel94** [comment]
Yeah, there is a bug in the gulp script, line 38 should be `return gulp.src("*.ts", { read: false })` - ts, not js 😉

**Comment by NoNameProvided** [comment]
Ok I have pushed it, but I am not sure how should I test this? All the other tests are checking for "success" scenarios, and I have to check for "fail" scenario, but as soon as an error happen the errorHandler takes it, so I cannot set any property or do anything else.

**Comment by 19majkel94** [comment]
>  but as soon as an error happen the errorHandler takes it, so I cannot set any property or do anything else

So describe the behaviour - throw e.g. `NotAcceptableError` in middleware and check if the response od the app is correct - status code 406 and the message body property is matching. No need to make a hacks and patching error handlers.

**Comment by NoNameProvided** [comment]
Ok, I did, and spotted an issue, its abou referencing `this`, while calling `this.handlerError` so I added the most ugliest solution to fix it for now. lol.

**Comment by NoNameProvided** [comment]
Done, and catched an other error thanks to the test

**Comment by 19majkel94** [comment]
> Ok, I did, and spotted an issue, its about referencing this, while calling this.handlerError so I added the most ugliest solution to fix it for now. lol.

I saw this but I'm glad you have remind about arrow functions :wink:

**Comment by NoNameProvided** [comment]
So is it good to go? Do we have an ETA for the next release? Actually we kind of need this feature at Monday 8am :D

File: src/driver/express/ExpressDriver.ts [hunk 1]
```diff
@@ -329,8 +329,20 @@ export class ExpressDriver extends BaseDriver implements Driver {
         const middlewareFunctions: Function[] = [];
         uses.forEach(use => {
             if (use.middleware.prototype && use.middleware.prototype.use) { // if this is function instance of MiddlewareInterface
-                middlewareFunctions.push(function (request: any, response: any, next: (err: any) => any) {
-                    return (getFromContainer(use.middleware) as ExpressMiddlewareInterface).use(request, response, next);
+                middlewareFunctions.push((request: any, response: any, next: (err: any) => any) =>{
```
Comment (pleerock) [inline src/driver/express/ExpressDriver.ts]: missing space in `=>{`

File: test/functional/express-middlewares.spec.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import { NotAcceptableError } from './../../src/http-error/NotAcceptableError';
+import { ExpressErrorMiddlewareInterface } from './../../src/driver/express/ExpressErrorMiddlewareInterface';
```
Comment (pleerock) [inline test/functional/express-middlewares.spec.ts]: please follow same syntax. use `"` and no spaces in import.

**Comment by NoNameProvided** [comment]
Fixed

File: test/functional/express-middlewares.spec.ts [hunk 3]
```diff
@@ -175,4 +203,12 @@ describe("express middlewares", () => {
             });
     });
 
+    it("should handle errors in custom middlewares", () => {
+        return chakram
+            .get("http://127.0.0.1:3001/customMiddlewareWichThrows")
```
Comment (pleerock) [inline test/functional/express-middlewares.spec.ts]: what about koa?

Comment (pleerock) [inline test/functional/express-middlewares.spec.ts]: how this new functionality stays with koa?

**Comment by NoNameProvided** [comment]
True, pushed tests for koa as well

**Comment by NoNameProvided** [comment]
Btw do I break the codecov tests or is it misconfigured?

**Comment by NoNameProvided** [comment]
@pleerock is it good to go?

**Comment by NoNameProvided** [comment]
@pleerock is there anything which prevents this pr from merging?

**Comment by pleerock** [comment]
thank you @NoNameProvided

**Pull request merged**